### PR TITLE
[UNR-4453] Implement Routing worker CrossServer RPCs

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Implementation of the reliable CrossServer RPCs using a routing worker.
Read in commit order to not be overwhelmed.

#### TLDR : 
- New RPC endpoints
- Launches an additional worker, integrated into the SpatialNetDriver, that only does routing work.
- Implementation selected through a setting (the legacy implementation is still accessible).

#### Overview of the system : 
Each actor has 4 new components. A sender, a receiver, and two other write the corresponding ACKs respectively.
The sender/receiver components are like regular ring buffer RPC components, with an additional array used to reference the target/sender of an RPC as well as it ID.
The ACK component is an array of ACK items which contains the sender and RPC ID. This is needed because the published RPCs can be received/acked in any order.

An actor has authority over its Sender and ReceiverACK component.
It is able to publish RPCs to its sender component to be executed on a target Actor.
It is able to acknowledge reception of RPCs sent from another Actor by publishing an ACK to its ReceiverACK component.

The routing worker is authoritative over all the Receiver and SenderACK components.
It transfers RPCs published on the Sender to the target Receiver.
When the Receiver has ACKed the RPC, its writes and ACK to the sender.
When the Sender has cleared the RPC from its components, the SenderACK and Receiver are cleared.

The routing worker runs as a new worker type. It does not execute any part of the UE stack, it just uses the SpatialNetDriver to get its connection to SpatialOS.

NB : Right now, overflowed RPCs are dropped. The overflow mechanism in the SpatialRPCService need a big refactor before that can be properly handled.

#### Release note
WIP

#### Tests
Added unit tests for the communication between CrossServerRPCService and SpatialRoutingSystem, and tested with the existing ServerToServer gym.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -368,78 +382,87 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 {
 	InitializeSpatialOutputDevice();
 
-	Dispatcher = MakeUnique<SpatialDispatcher>();
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:371]: This diff really boils down to putting the core class initialization behind `if (WorkerType == SpatialConstants::DefaultServerWorkerType || WorkerType == SpatialConstants::DefaultClientWorkerType)`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 2]
```diff
@@ -2590,39 +2803,78 @@ void USpatialNetDriver::TryFinishStartup()
 
 	if (IsServer())
 	{
-		if (!PackageMap->IsEntityPoolReady())
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2593]: This diff boils down to putting the regular server startup sequence behind
`if (WorkerType == SpatialConstants::DefaultServerWorkerType)`

File: SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace SpatialGDK
+{
+CrossServerEndpoint::CrossServerEndpoint(Schema_ComponentData* Data)
+	: ReliableRPCBuffer(ERPCType::CrossServer)
+{
+	ReadFromSchema(Schema_GetComponentDataFields(Data));
+}
+
+void CrossServerEndpoint::ApplyComponentUpdate(Schema_ComponentUpdate* Update)
```
Comment (jacquese) [inline SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp:13]: would appreciate comments throughout this file describing what this is doing- or at least a pointer to where it is described. For example, it's not immediately clear why the clear field logic in this function is the way that it is.

**Review by jacquese — COMMENTED** [review]
reviewing commit by commit to hopefully get feedback out quicker. Counterpart- is that the other entity in the cross server RPC?

File: SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp (hunk 3 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp:13]: Right, that's me enforcing that both fields (which should be valid/invalid together) are cleared at the same time. But I'm just overzealous.

**Comment by ImprobableNic** [comment]
Yes, the counterpart is the other entity in a sender/receiver scenario.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -112,6 +112,19 @@ USpatialNetDriver::~USpatialNetDriver() = default;
...
 {
+	if (!bConnectAsClient)
+	{
+		USpatialGameInstance* GameInstance = GetGameInstance();
+
+		if (GameInstance != nullptr)
+		{
+			if (GameInstance->GetSpatialWorkerType() == SpatialConstants::RoutingWorkerType)
+			{
+				NetServerMaxTickRate = 120;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:123]: I wonder how heavy a UE instance for this purpose is (and what the cost to users is). Perhaps a future possibility is a tiny c worker that runs on the runtime node and the cost can be considered an extension of runtime provided functionality (and avoid interfering with bin-packing)

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -913,6 +936,14 @@ void USpatialNetDriver::BeginDestroy()
 			}
 		}
 
+		if (RoutingSystem)
+		{
+			RoutingSystem->Destroy(Connection);
+
+			Connection->Flush();
+			FPlatformProcess::Sleep(0.1f);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:944]: What are these sleeps for?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -1613,6 +1644,8 @@ void USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConne
 
 #endif // WITH_SERVER_CODE
 
+thread_local TArray<AActor*> GSenderStack;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1647]: Is there a reason this isn't a member variable? In the editor I think all NetDrivers will run in serial on the same thread

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -1646,14 +1679,61 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
...
+			return CrossServerRPCSender->SendCommand(MoveTemp(CallingObjectRef), CallingObject, Function, MoveTemp(Payload), Info, {});
+		}
+		else
+		{
+			AActor* SenderActor = nullptr;
+
+			if (GSenderStack.Num() == 0)
+			{
+				// These will eventually become error cases
+				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Missing sender Actor to call RPC %s on target %s"), *Function->GetName(),
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1699]: Is it worth adding something here in the interim so we don't forget?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 8]
```diff
@@ -1646,14 +1679,61 @@ void USpatialNetDriver::ProcessRPC(AActor* Actor, UObject* SubObject, UFunction*
...
+			}
+			else
+			{
+				SenderActor = GSenderStack.Last();
+			}
+
+			if (SenderActor == nullptr)
+			{
+				// These will eventually become error cases
+				// UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Null sender Actor set to call RPC %s on target %s"), *Function->GetName(),
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1711]: Same here ^

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -1836,45 +1916,63 @@ void USpatialNetDriver::TickDispatch(float DeltaTime)
...
+				RPCService->AdvanceView();
+			}
 
-		{
-			SCOPE_CYCLE_COUNTER(STAT_SpatialProcessOps);
-			Dispatcher->ProcessOps(GetOpsFromEntityDeltas(Connection->GetEntityDeltas()));
-			Dispatcher->ProcessOps(Connection->GetWorkerMessages());
-			CrossServerRPCHandler->ProcessMessages(Connection->GetWorkerMessages(), DeltaTime);
-		}
+			if (ActorSystem.IsValid())
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1950]: Some of these systems aren't initialized for the routing worker - maybe this path can be branchless with .IsValid checks

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1950]: Don't know if separating provides more context though, so just an observation

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 10]
```diff
@@ -1836,45 +1916,63 @@ void USpatialNetDriver::TickDispatch(float DeltaTime)
...
+			}
+
+			if (WellKnownEntitySystem.IsValid())
+			{
+				WellKnownEntitySystem->Advance();
+			}
 		}
 
-		if (WellKnownEntitySystem.IsValid())
+		if (RoutingSystem)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1973]:
```suggestion
		if (RoutingSystem != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 9 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1950]: Or maybe the logic could be seperated into something like TickDispatchRoutingWorker and TickDispatchServerClientWorker?

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1950]: Or a separate NetDriver? I don't know how much is actually shared though :D

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 11]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
+				for (uint32_t i = 0; i < Op.result_count; ++i)
+				{
+					const Worker_Entity& Entity = Op.results[i];
+					for (uint32_t j = 0; j < Entity.component_count; ++j)
+					{
+						const Worker_ComponentData& ComponentData = Entity.components[j];
+						if (ComponentData.component_id == SpatialConstants::WORKER_COMPONENT_ID)
+						{
+							Schema_Object* Fields = Schema_GetComponentDataFields(ComponentData.schema_type);
+							FString WorkerId = SpatialGDK::GetStringFromSchema(Fields, 1);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2842]: Might be worth using a SpatialConstant for the field ids

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 12]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
+							FString WorkerType = SpatialGDK::GetStringFromSchema(Fields, 2);
+							if (WorkerType == SpatialConstants::RoutingWorkerType.ToString())
+							{
+								NetDriver->RoutingWorkerId = Entity.entity_id;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2846]: Is ID=0 a valid entity ID? Could end up recursively in here if so

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 13]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
+				}
+			}
+			NetDriver->bRoutingWorkerQueryInFlight = false;
+		});
+		Receiver->AddEntityQueryDelegate(RequestID, RoutingWorkerQueryDelegate);
+	}
+	else if (RoutingPartition == 0)
+	{
+		Worker_ComponentConstraint ComponentConstraint{};
+		ComponentConstraint.component_id = 66;
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2860]:
```suggestion
		ComponentConstraint.component_id = SpatialConstants::ROUTING_WORKER_COMPONENT_ID;
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 14]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
+			USpatialNetDriver* NetDriver = WeakNetDriver.Get();
+			if (Op.status_code == WORKER_STATUS_CODE_SUCCESS)
+			{
+				for (uint32_t i = 0; i < Op.result_count; ++i)
+				{
+					const Worker_Entity& Entity = Op.results[i];
+					for (uint32_t j = 0; j < Entity.component_count; ++j)
+					{
+						const Worker_ComponentData& ComponentData = Entity.components[j];
+						if (ComponentData.component_id == 66)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2889]:
```suggestion
						if (ComponentData.component_id == SpatialConstants::ROUTING_WORKER_PARTITION_COMPONENT_ID)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 15]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
+		TWeakObjectPtr<USpatialNetDriver> WeakNetDriver(this);
+		EntityQueryDelegate RoutingWorkerQueryDelegate;
+		RoutingWorkerQueryDelegate.BindLambda([WeakNetDriver](const Worker_EntityQueryResponseOp& Op) {
+			if (!WeakNetDriver.IsValid())
+			{
+				return;
+			}
+
+			USpatialNetDriver* NetDriver = WeakNetDriver.Get();
+			if (Op.status_code == WORKER_STATUS_CODE_SUCCESS)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2881]: Perhaps some error/logging if the code != SUCCESS?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 16]
```diff
@@ -2691,6 +2796,114 @@ void USpatialNetDriver::DelayedRetireEntity(Worker_EntityId EntityId, float Dela
...
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2831]: Perhaps some error/logging if the code != SUCCESS?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 17]
```diff
@@ -2699,47 +2912,73 @@ void USpatialNetDriver::TryFinishStartup()
...
-		else
+		USpatialGameInstance* GameInstance = GetGameInstance();
+		FName WorkerType = GameInstance->GetSpatialWorkerType();
+
+		if (WorkerType == SpatialConstants::RoutingWorkerType)
 		{
-			UE_LOG(LogSpatialOSNetDriver, Log, TEXT("Ready to begin processing."));
+			// RoutingWorkerId = Connection->GetWorkerId();
+
+			SpatialGDK::FSubView& newView =
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2922]:
```suggestion
			SpatialGDK::FSubView& NewView =
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+		case EntityDelta::REMOVE:
+		case EntityDelta::TEMPORARILY_REMOVED:
+			CrossServerDataStore.Remove(Delta.EntityId);
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID));
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_RECEIVER_ACK_ENDPOINT_COMPONENT_ID));
+			if (Delta.Type == EntityDelta::TEMPORARILY_REMOVED)
+			{
+				PopulateDataStore(Delta.EntityId);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:105]: Does TEMPORARILY_REMOVED mean the data won't be in the subview?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp [hunk 19]
```diff
@@ -0,0 +1,476 @@
...
+			break;
+		}
+	}
+}
+
+void CrossServerRPCService::EntityAdded(const Worker_EntityId EntityId)
+{
+	for (const Worker_ComponentId ComponentId : SubView->GetView()[EntityId].Authority)
+	{
+		if (!IsCrossServerEndpoint(ComponentId))
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:148]: The authority table contains component set ids iirc (there was a similar fix required for on-add rpc execution in the client/server service)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp [hunk 20]
```diff
@@ -0,0 +1,476 @@
...
+	RPCRingBufferDescriptor Descriptor = RPCRingBufferUtils::GetRingBufferDescriptor(ERPCType::CrossServer);
+	uint32 Field = Descriptor.GetRingBufferElementFieldId(ERPCType::CrossServer, SlotIdx + 1);
+
+	Schema_Object* RPCObject = Schema_AddObject(EndpointObject, Field);
+
+	RPCTarget Target(CrossServerRPCInfo(EntityId, NewRPCId));
+	CrossServer::WritePayloadAndCounterpart(EndpointObject, Payload.Payload, Target, SlotIdx);
+
+	Schema_ClearField(EndpointObject, Descriptor.LastSentRPCFieldId);
+	Schema_AddUint64(EndpointObject, Descriptor.LastSentRPCFieldId, SenderState.LastSentRPCId);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:65]: Does this make the clear above it redundant?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp [hunk 21]
```diff
@@ -0,0 +1,476 @@
...
+		Endpoints = &CrossServerDataStore.Add(Sender.Entity);
+	}
+	else
+	{
+		EndpointObject = Schema_GetComponentUpdateFields(RPCStore->GetOrCreateComponentUpdate(SenderEndpointId));
+	}
+
+	CrossServer::WriterState& SenderState = Endpoints->SenderState;
+
+	TOptional<uint32> Slot = SenderState.Alloc.PeekFreeSlot();
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:47]: Is it worth just attempting to reserve the slot here? To avoid potential API misuse by using a peeked slot without committing it. Perhaps it could safeguarded a little by ensuring that Reserve is followed by Commit and not Reserve again (in the case of an early out or something)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 22]
```diff
@@ -234,6 +235,9 @@ void USpatialSender::CreateServerWorkerEntity()
 void USpatialSender::RetryServerWorkerEntityCreation(Worker_EntityId EntityId, int AttemptCounter)
 {
 	check(NetDriver != nullptr);
+	const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>();
+
+	if (Settings->CrossServerRPCImplementation == ECrossServerRPCImplementation::RoutingWorker) {}
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:240]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 23]
```diff
@@ -687,18 +711,36 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
+										 const FUnrealObjectRef& TargetObjectRef)
+{
+	const FRPCInfo& RPCInfo = ClassInfoManager->GetRPCInfo(TargetObject, Function);
+	const EPushRPCResult Result =
+		RPCService->PushRPC(TargetObjectRef.Entity, Sender, RPCInfo.Type, Payload, Channel->bCreatedEntity, TargetObject, Function);
+
+	// if (Result == EPushRPCResult::Success)
+	//{
+	//	FlushRPCService();
+	//}
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:742]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+{
+	ReadFromSchema(Schema_GetComponentUpdateFields(Update));
+	uint32 ClearCount = Schema_GetComponentUpdateClearedFieldCount(Update);
+	TArray<Schema_FieldId> ClearedFields;
+	ClearedFields.SetNum(ClearCount);
+	Schema_GetComponentUpdateClearedFieldList(Update, ClearedFields.GetData());
+
+	for (auto Field : ClearedFields)
+	{
+		if (Field >= 1)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp:52]:
```suggestion
		if (ensure(Field >= 1))
```

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Schema/CrossServerEndpoint.cpp:52]: Should this ever happen?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialRoutingSystem.h [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+#include "SpatialView/SubView.h"
+#include "Utils/CrossServerUtils.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogSpatialRoutingSystem, Log, All)
+
+class SpatialOSWorkerInterface;
+
+namespace SpatialGDK
+{
+struct SpatialRoutingSystem
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialRoutingSystem.h:17]: Nit: Seems more like class behavior

**Comment by improbable-danny** [comment]
Done a quick pass, overall looks good but I don't have a lot of context on the details

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 6 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1647]: That might be true, but I do not want to take any risk.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 5 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:944]: Don't know, I reproduced the pattern, just in case.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 4 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:123]: Right now the impact has not been measured. The instance would likely be quite big since it would load the map as well, so that the first optim to consider.
Having a small C worker would likely be much better, but we do not have the infra to launch utility workers right now.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 12 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2846]: I'm pretty sure it is not.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp (hunk 18 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:105]: It means that the entity flashed out/in interest, so it disappeared for a while, then reappeared.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 23 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:742]: That's right, it should be removed, but that could be considered as a controversial change (more latency), so I'm leaving it there for discussion.
But the crux of it is that if you ever sent a RPC alongside modifying your properties, making the remote side receive the RPC before you had the opportunity of replicating your properties as well might cause issues and unexpected/broken invariants. So that change is likely the correct thing to do.

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:742]: Is this not also the case for native? Commands are read in-order and RPCs are executed as the stream is read iirc - so properties will be applied after RPCs (not 100% on this)

Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:742]: Ah interesting, I think the case of native is that this does happen in a defined way - DataChannel.cpp reads and actions items in order (RPCs are written immediately and properties at end-of-frame) so it seems RPCs should always happen first (chekced with some breakpoints also). Although the behavior is slightly more defined in that RPCs then properties are executed. Unmapped RPC args are sometimes deferred until after property updates though which resolves some of that ambiguity (FObjectReplicator::ReceivedBunch bCanDelayRPCs)

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1950]: They are separated though. May be hard to read because the diff is bad, but this gated by if (bIsDefaultServerOrClientWorker)
I do agree though that this can be branchless because most of these will be initialized all the time, but that's another debate.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp (hunk 19 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:148]: Nice catch!

File: SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp (hunk 21 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp:47]: True, guess the code changed as to make it not necessary to separate them.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 23 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp:742]: Reading more on this, seems like by default, everything is sent immediately, except unreliable net multicast RPCs.
So it looks like native UE is not concerned about replication frame consistency.

**Review by samiwh — APPROVED** [review]
lgtm
Some things to note:
Currently the lack of blank entities means it's impossible to deal with case of rpcs being sent to actors that aren't fully initialised in an ideal way.

The spatial view test helpers don't exist to make that work as we'd like.

These are likely to be resolved as the current refactoring work continues. Until then it's agreed it's better to have this merged in and maintained instead.
